File: exercises/f1c8ce22-90d4-4feb-b99a-8acc686c7496/solutions/d0c09c4b-bcbf-437a-8004-c25176cda6b5/Chemistry.java
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.PrintStream;

/**
 * @author Margarida Mamede
 */


public class Chemistry
{

    private BufferedReader formula;

    private char symbol;


    public Chemistry( BufferedReader theFormula )
    {
        formula = theFormula;
    }


    private void writeAndGoForward( PrintStream out ) throws IOException
    {
        out.print(symbol);
        symbol = (char) formula.read();
    }


    public void decompose( PrintStream out ) throws IOException
    {
        symbol = (char) formula.read();
        do 
            {
                // An element begins with an upper-case letter.
                this.writeAndGoForward(out);

                // There may be lower-case letters.
                while ( symbol >= 'a' && symbol <= 'z' )
                    this.writeAndGoForward(out);
                    
                out.print(" ");  

                if ( symbol >= '1' && symbol <= '9' )
                    // The number of atoms is given.
                    do
                        this.writeAndGoForward(out);
                    while ( symbol >= '0' && symbol <= '9' );
                else
                    // The number of atoms is 1.
                    out.print(1);  

                out.println();  
            }
        while ( symbol != '.' );
    }


    public static void main( String[] args ) throws IOException
    {
        BufferedReader input = 
            new BufferedReader( new InputStreamReader(System.in) );

        new Chemistry(input).decompose(System.out);

        input.close();
    }


}

File: exercises/f1c8ce22-90d4-4feb-b99a-8acc686c7496/solutions/91a2b3d8-73ce-4b21-8658-86714b2dc63d/chemistry.c
#include<stdio.h>


/*
 * @author Margarida Mamede
 */


void writeAndGoForward( char *character )
{
    printf("%c", *character);
    scanf("%c", character);
}


void decompose( )
{
    char symbol;

    scanf("%c", &symbol);
    do 
        {
	    // An element begins with an upper-case letter.
	    writeAndGoForward(&symbol);

	    // There may be lower-case letters.
	    while ( symbol >= 'a' && symbol <= 'z' )
	        writeAndGoForward(&symbol);
                    
	    printf(" ");  

	    if ( symbol >= '1' && symbol <= '9' )
	        // The number of atoms is given.
	        do
		    writeAndGoForward(&symbol);
		while ( symbol >= '0' && symbol <= '9' );
	    else
	        // The number of atoms is 1.
                printf("1");  

	    printf("\n");  
	}
    while ( symbol != '.' );
}


int main( )
{
    decompose();
    return 0;
}
